autogpt_platform/backend/backend/integrations/oauth/google.py

```diff
@@ -35,36 +47,74 @@ def get_login_url(self, scopes: list[str], state: str) -> str:
         )
         return authorization_url
 
-    def exchange_code_for_tokens(self, code: str) -> OAuth2Credentials:
-        flow = self._setup_oauth_flow(None)
+    def exchange_code_for_tokens(
+        self, code: str, scopes: list[str]
+    ) -> OAuth2Credentials:
+        logger.debug(f"Exchanging code for tokens with scopes: {scopes}")
+
+        # Use the scopes from the initial request
+        flow = self._setup_oauth_flow(scopes)
         flow.redirect_uri = self.redirect_uri
-        flow.fetch_token(code=code)
 
-        google_creds = flow.credentials
-        username = self._request_email(google_creds)
+        logger.debug("Fetching token from Google")
 
-        # Google's OAuth library is poorly typed so we need some of these:
-        assert google_creds.token
-        assert google_creds.refresh_token
-        assert google_creds.expiry
-        assert google_creds.scopes
-        return OAuth2Credentials(
-            provider=self.PROVIDER_NAME,
-            title=None,
-            username=username,
-            access_token=SecretStr(google_creds.token),
-            refresh_token=SecretStr(google_creds.refresh_token),
-            access_token_expires_at=int(google_creds.expiry.timestamp()),
-            refresh_token_expires_at=None,
-            scopes=google_creds.scopes,
-        )
+        try:
+            # Disable scope check in fetch_token
+            flow.oauth2session.scope = None
+            token = flow.fetch_token(code=code)
+            logger.debug("Token fetched successfully")
+
+            # Get the actual scopes granted by Google
+            granted_scopes = token.get("scope", [])
+            if isinstance(granted_scopes, str):
+                granted_scopes = granted_scopes.split()
+            logger.debug(f"Scopes granted by Google: {granted_scopes}")
+
+            google_creds = flow.credentials
+            logger.debug(f"Received credentials: {google_creds}")
+
+            logger.debug("Requesting user email")
+            username = self._request_email(google_creds)
+            logger.debug(f"User email retrieved: {username}")
+
+            # Create OAuth2Credentials with the granted scopes
+            credentials = OAuth2Credentials(
+                provider=self.PROVIDER_NAME,
+                title="Google Account",
+                username=username,
+                access_token=SecretStr(google_creds.token or ""),
+                refresh_token=(
+                    SecretStr(google_creds.refresh_token)
+                    if google_creds.refresh_token
+                    else None
+                ),
```

Comment:
do we have a better way to handle refreshing tokens?

Reply:
Why/when would there not be a refresh token?

Reply:
lol there should be... fixed